packages/cli-repl/src/async-repl.ts

```diff
@@ -0,0 +1,131 @@
+              reject(new Error('Asynchronous execution was interrupted by `SIGINT`'));
+            };
+            previousSigintListeners = repl.rawListeners('SIGINT');
+
+            repl.removeAllListeners('SIGINT');
+            repl.once('SIGINT', sigintListener);
+          }
+
+          const evalResult = asyncEval(originalEval, input, context, filename);
+
+          if (sigintListener !== undefined) {
+            process.once('SIGINT', sigintListener);
+          }
+          evalResult.then(resolve, reject);
+        });
+      } finally {
+        // Remove our 'SIGINT' listener and re-install the REPL one(s).
+        if (sigintListener !== undefined) {
+          repl.removeListener('SIGINT', sigintListener);
+          process.removeListener('SIGINT', sigintListener);
+        }
+        for (const listener of previousSigintListeners) {
+          repl.on('SIGINT', listener);
+        }
+      }
+    } catch (err) {
+      try {
+        if (isRecoverableError(input)) {
+          return callback(new Recoverable(err));
+        }
+        return callback(err);
+      } catch (callbackErr) {
+        return callback(wrapCallbackError(callbackErr));
+      }
+    }
+    try {
+      return callback(null, result);
+    } catch (callbackErr) {
+      return callback(wrapCallbackError(callbackErr));
+    }
+  };
+
+  return repl;
+}
+
+function wrapNoSyncDomainError<Args extends any[], Ret>(fn: (...args: Args) => Ret) {
+  return (...args: Args): Ret => {
+    const origEmit = Domain.prototype.emit;
+
+    // When the Node.js core REPL encounters an exception during synchronous
+    // evaluation, it does not pass the exception value to the callback
+    // (or in this case, reject the Promise here), as one might inspect.
+    // Instead, it skips straight ahead to abandoning evaluation and acts
+    // as if the error had been thrown asynchronously. This works for them,
+    // but for us that's not great, because we rely on the core eval function
+    // calling its callback in order to be informed about a possible error
+    // that occurred (... and in order for this async function to finish at all.)
+    // We monkey-patch `process.domain.emit()` to avoid that, and instead
+    // handle a possible error ourselves:
+    // https://github.com/nodejs/node/blob/59ca56eddefc78bab87d7e8e074b3af843ab1bc3/lib/repl.js#L488-L493
+    // It's not clear why this is done this way in Node.js, however,
+    // removing the linked code does lead to failures in the Node.js test
+    // suite, so somebody sufficiently motivated could probably find out.
+    // For now, this is a hack and probably not considered officially
+    // supported, but it works.
+    // We *may* want to consider not relying on the built-in eval function
+    // at all at some point.
+    Domain.prototype.emit = function(ev: string, ...eventArgs: any[]): boolean {
+      if (ev === 'error') {
+        this.exit();
```

Comment:
This line was missing and causing Node.js core deprecation warnings to be emitted that some of us noticed, yay for testing to catch this :)

Reply:
Nice!!